chromium-content-api/preload-chromium-content-api.c

```diff
@@ -0,0 +1,151 @@
+/* Copyright (C) 2020 Canonical, Ltd.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; version 3.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef _GNU_SOURCE
+#define _GNU_SOURCE
+#endif
+
+#define __USE_GNU
+
+#include <dlfcn.h>
+#include <fcntl.h>
+#include <linux/limits.h>
+#include <stdarg.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <unistd.h>
+
+static int (*original_open)(const char *, int, ...);
+static int (*original_unlink)(const char *);
+
+
+void debug(char *fmt, ...)
+{
+	if (secure_getenv("SHM_CHROMIUM_DEBUG")) {
+		va_list va;
+		va_start(va, fmt);
+		fprintf(stderr, "SHM_CHROMIUM_WRAP: ");
+		vfprintf(stderr, fmt, va);
+		va_end(va);
+		fprintf(stderr, "\n");
+	}
+}
+
+const char *get_snap_name(void)
+{
+	const char *snapname = getenv("SNAP_INSTANCE_NAME");
+	if (snapname == NULL) {
+		snapname = getenv("SNAP_NAME");
+	}
+	if (snapname == NULL) {
+		debug("SNAP_NAME and SNAP_INSTANCE_NAME not set");
+	}
+	return snapname;
+}
+
+
+static char * redirect_shm(const char *snapname, const char *pathname)
+{
+	size_t max_len = PATH_MAX + 1;
+	/* New path includes snap.<SNAP_INSTANCE_NAME> and NULL.  == 6 + 1 */
```

Comment:
```suggestion:
	/* Rename: /dev/shm/shm-<remainder> -> /dev/shm/snap.<snap-name>.shm-<remainder> */
```
Looking at this again, I wonder if it should just be:

Rename: `/dev/shm/<shm-file>` => `/dev/shm/snap.<snap-name>.<shm-file>`

Any reason for filtering for "shm-"?  I know chromium will use that prefix, but wonder if there's a reason to do that.
In this case, we could (safely?) rely on `basename()` rather than sscanf.

Would the rules allow directories under /dev/shm/snap.snap-name.shm/ ?

Reply:
@jdstrand ?

Reply:
We discussed on mattermost and I agree that snapcraft-preloads should support all of /dev/shm like @cjp256 suggests (this is what snapcraft-preload does today). Now, whether you want to have one for all and another for just chromium, or just the one for all is up to you.